### ISSUE TITLE
Interpolation from (and to) irregular grid

### ISSUE THREAD

**Issue by Balinus** [juliohm/GeoStats.jl#22]
Hello,  

I'm looking to a solution to the following problem. With climate model data, the grids are often "curvilinear" (e.g. rotated poles, lambert conformal conic , etc...). I'd like to interpolate from one grid to another. I have access to the longitude and latitude grids (i.e. 2D array for longitude "LON" and 2D array for latitude "LAT") for both the source and destination grid. Hence, the data values coordinates `[i, j]` are specified by `LON[i, j]` and `LAT[i, j]`.

My aim is to simply do an inverse weight distance for each points of the destination grid, using the 2D array of longitude and latitude. However, when I look at the codes, the "x" and "y" are vectors and not array. I was wondering if it's possible to use GeoStats with such 2D coordinates arrays. Perhaps it's simply a matter of unfolding the arrays to 1D vectors?

Thanks for any help!

**Comment by Balinus** [comment]
Thanks!

I'll see what I can do!

**Comment by juliohm** [comment]
The spatial data interface should take a couple of minutes to implement. The last method `view` is not strictly necessary, only if you want to perform cross-validation with other estimation methods later on.

The spatial domain interface is the one where optimizations can help. You can start with a very simple implementation that does what you want and then optimize it with more time.

To define these types on your scripts, you can do something like:

```julia
using GeoStats

struct CurvilinearGridData{T<:Real,N} <: AbstractSpatialData
  LAT::Matrix{Float64}
  LON::Matrix{Float64}
  vals::Matrix{Float64}
end

# define the methods in the interface...
```

**Comment by Balinus** [comment]
Thanks for the hints! :)

**Comment by juliohm** [comment]
I was thinking about this issue yesterday, and I think the solution here would be to add two types `StructuredData` and `StructuredGrid` that are constructed based on the NetCDF convention of LAT, LON, ..., and VALUES arrays.

Since the data already comes in a grid format, I am planning to construct a lattice graph internally to look up things efficiently.

**Comment by juliohm** [comment]
We have `StructuredGridData` implemented already, now we only need to implement `StructuredGrid`.

**Comment by juliohm** [comment]
As we were discussing on Gitter, the issue is theoretically solved. We have `StructuredGrid` and `StructuredGridData` implemented :+1: 

Below are examples of `StructuredGrid`, we still need better plots for `StructuredGridData`:

```julia
using GeoStats
using Plots

# create some 2D coordinates
nx, ny = 20, 10
X = [x for x in range(0,10,length=nx), j in 1:ny]
Y = sin.(X) .+ [0.5j for i in 1:nx, j in 1:ny]

g = StructuredGrid(X, Y)

plot(g)
```
![Screenshot from 2019-05-19 18-31-53](https://user-images.githubusercontent.com/3345261/57988456-1559c100-7a65-11e9-9bd6-f889196e3498.png)

```julia
# create some 3D coordinates
nx, ny, nz = 20, 10, 10
X = [x for x in range(0,10,length=nx), j in 1:ny, k in 1:nz]
Y = sin.(X) .+ [0.5j for i in 1:nx, j in 1:ny, k in 1:nz]
Z = [1.0(k-1) for i in 1:nx, j in 1:ny, k in 1:nz]

g = StructuredGrid(X, Y, Z)

plot(g, camera=(30,60), size=(600,600))
```
![Screenshot from 2019-05-19 18-32-33](https://user-images.githubusercontent.com/3345261/57988464-3e7a5180-7a65-11e9-9e65-61b2f1e3ba86.png)

**Comment by Balinus** [comment]
That's awesome! Thanks!

**Comment by briochemc** [comment]
Sorry for digging this up, but I am guessing there is a solution to the interpolation question since this was closed? Could you provide a MWE of interpolation from one grid to another here? (A search for "curvilinear" in JuliaEarth only shows this issue.)

**Comment by juliohm** [comment]
Hi @briochemc , we are still working on curvilinear meshes in Meshes.jl. We will get there at some point. 🙏🏽 

Can you comment about your use case to see how it fits with the roadmap?

**Comment by briochemc** [comment]
Well I think it's very similar to Balinus' case... Climate models (ocean models for me) have their own (sometimes curvilinear) grids and it would be good to have a robust tool/API to interpolate data between them to compare the same data from different models for example

I actually also opened [this issue with Interpolations.jl](https://github.com/JuliaMath/Interpolations.jl/issues/432) because I thought this might be a more general task/feature.

My personal use case of the moment is that I have three-dimensional CESM model output data embedded in a curvilinear grid for lat/lon and I want to interpolate that data onto a simpler rectangular grid to work with it.

**Comment by juliohm** [comment]
The example helped understand what your goal is. Let me clarify what we can do currently and why I mentioned it is still not ideal.

I will assume that you have a collection of corner coordinates (LAT, LON) and that you want to interpolate values on the grid elements (quadrangles). You can create a vector of `Meshe.Point` with the lat, lon values for each vertex in the grid and save in a vector `points`. You can then create a `mesh = SimpleMesh(points, GridTopology(19, 39))` and proceed with the GeoStats.jl pipeline as usual, and do interpolation, simulation, etc. The problem is the following:

1. Currently we do not have a representation for curve line segments. So the dots will be connected by straight lines like the plot you shared in the other issue.
2. We do not consider the shapes of the quadrangles correctly yet, and methods such as Kriging rely on integrals over shapes to regularize the correlation function.

So in summary, you can create a `SimpleMesh` and do interpolation, just the result won't be the most mathematically correct result possible yet.

**Comment by briochemc** [comment]
I think there may have been a slight confusion in how I phrased my example. What I have is all the lat/lon coordinates of the grid vertices and a variable that has a value at each vertex. So three 2D arrays (`LAT`, `LON`, and `V`) like in the Interpolations issue posted. And I would like to interpolate the values `V` to any other lat/lon that is not necessarily an element of the grid. Is that possible currently? (The next goal being to interpolate `V` at all the points of another distince grid.)

Not being perfectly accurate is fine (an error within 1% would probably go unnoticed to me frankly). I could go with nearest neighbor value but I'm sure I can do better with your guys' tools 😄

**Comment by juliohm** [comment]
What you can do currently in this case is simply create a PointSet with the
coordinates of the vertices of the grid and then another PointSet with the
target locations. This will generate a point interpolation and you could
use the Haversine as usual to compute the distance between the points in
Kriging, IDW, LWR, etc.

We are still polishing the curvilinear grid api to facilitate the
construction of objects such as SimpleMesh(points, GridTopology(...)) which
is too low-level.  We also want to provide the interpolation in the
aerial elements at some point, it is in the roadmap :)

Em seg., 21 de jun. de 2021 às 09:56, Benoit Pasquier <
***@***.***> escreveu:

**Comment by briochemc** [comment]
I'm still a bit confused about how to do that... Would you mind giving me a little bit more hand-holding? Here is a MWE with 2 `PointSet`s. How do I do the interpolation at the end?

<img width="584" alt="Screen Shot 2021-06-22 at 11 21 37 am" src="https://user-images.githubusercontent.com/4486578/122848069-2d8b2100-d34c-11eb-8bc9-51f91d191075.png">


```julia
# The curvilinear grid:
_x = -20:20
_y = -10:10
LON = [x + 10cos(y/(40+x)) for x in _x, y in _y]
LAT = [y + 0.001x^3 for x in _x, y in _y]
# The target grid (rectangular in this case)
_x2 = -10:5:35
_y2 = -20:5:10
LON2 = [x for x in _x2, y in _y2]
LAT2 = [y for x in _x2, y in _y2]
# The data on that grid:
f(lon, lat) = cosd(2lon) + sind(2lat) # some "true" function of lat/lon
V = [f(x,y) for (x,y) in zip(LON,LAT)] # values on curvilinear grid
# A plot to illustrate:
using Plots
lonrows = vcat(([x; NaN] for x in eachrow(LON))...)
loncols = vcat(([x; NaN] for x in eachcol(LON))...)
latrows = vcat(([x; NaN] for x in eachrow(LAT))...)
latcols = vcat(([x; NaN] for x in eachcol(LAT))...)
plot([lonrows; loncols], [latrows; latcols], lab="curvilinear grid", color=:black)
scatter!(LON[:], LAT[:], zcolor=V[:])
lonrows2 = vcat(([x; NaN] for x in eachrow(LON2))...)
loncols2 = vcat(([x; NaN] for x in eachcol(LON2))...)
latrows2 = vcat(([x; NaN] for x in eachrow(LAT2))...)
latcols2 = vcat(([x; NaN] for x in eachcol(LAT2))...)
plot!([lonrows2; loncols2], [latrows2; latcols2], lab="target grid", color=:red)

# Now to interpolate I make a PointSet of the original and target grids
using GeoStats # v0.25.2
PTS = PointSet([(x,y) for (x,y) in zip(LON[:],LAT[:])])
PTS2 = PointSet([(x,y) for (x,y) in zip(LON2[:],LAT2[:])])
# Next step is to "interpolate" onto values V onto the PTS2 point set... Help here! 
```

**Comment by juliohm** [comment]
Here is an estimation problem on the two point sets:

```julia
# fix coordinates of PTS2 to Float64
PTS2 = PointSet([Point2(x,y) for (x,y) in zip(LON2[:],LAT2[:])])

# georeference the data (columns of a table) on first point set
data = georef((V = V[:],), PTS)

# second point set is the domain of interpolation
domain = PTS2

# define estimation problem
problem = EstimationProblem(data, domain, :V)

plot(problem)
```
![image](https://user-images.githubusercontent.com/3345261/122919225-5ef8f200-d336-11eb-80d3-e0bfd956c972.png)

Now you can use any of the estimation solvers to solve the problem. Please let me know if something is not clear, we are working on improving the documentation to clarify this domain vs. data issue that many people get confused with. Also feel free to continue the discussion in Gitter or Zulip 👍🏽

**Comment by briochemc** [comment]
OK I think I got it. Thank you for helping! For the record, here is a MWE:

```julia
_x = -20:20
_y = -10:10
LON = [x + 10cos(y/(40+x)) for x in _x, y in _y]
LAT = [y + 0.001x^3 for x in _x, y in _y]
# The target grid (rectangular in this case)
_x2 = -10:5:35
_y2 = -20:5:10
LON2 = [x for x in _x2, y in _y2]
LAT2 = [y for x in _x2, y in _y2]
# The data on that grid:
f(lon, lat) = cosd(2lon) + sind(2lat) # some "true" function of lat/lon
V = [f(x,y) for (x,y) in zip(LON,LAT)] # values on curvilinear grid

using GeoStats # v0.25.2
origin = PointSet([Point2(x,y) for (x,y) in zip(LON[:],LAT[:])])
target = PointSet([Point2(x,y) for (x,y) in zip(LON2[:],LAT2[:])])
# georeference the data (columns of a table) on first point set
dat = georef((V = V[:],), origin)
# Make estimation problem
prob = EstimationProblem(dat, target, :V)
# Chose algorithm
algo = Kriging(:V => (distance=Haversine(), maxneighbors=10))
# Solve the estimation problem
sol = solve(prob, algo)
# Plot it
using Plots
plot(dat, lab="original data", legend=:topleft)
plot!(domain(sol), sol.V, marker=:square, lab="interpolated data")
```

that gives
<img width="584" alt="Screen Shot 2021-06-23 at 11 09 36 am" src="https://user-images.githubusercontent.com/4486578/123019287-8ec9f780-d413-11eb-99bc-9db117805dd7.png">
